packages/mongodb-ts-autocomplete/src/index.ts

```diff
@@ -0,0 +1,215 @@
+    // names from a listDatabases() or it could just be all the ones we've
+    // auto-completed for. The schemas can come from the autocompletion context.
+    // This can probably be added to the autocompletion context.
+    this.databaseNames = new Set();
+    this.databaseSchemas = Object.create(null);
+  }
+
+  addDatabase(databaseName: string) {
+    this.databaseNames.add(databaseName);
+    if (!this.databaseSchemas[databaseName]) {
+      this.databaseSchemas[databaseName] = new DatabaseSchema();
+    }
+  }
+
+  setDatabaseCollectionNames(databaseName: string, collectionNames: string[]) {
+    this.addDatabase(databaseName);
+    this.databaseSchemas[databaseName].setCollectionNames(collectionNames);
+  }
+
+  addCollectionSchema(
+    databaseName: string,
+    collectionName: string,
+    collectionSchema: JSONSchema
+  ) {
+    this.addDatabase(databaseName);
+    this.databaseSchemas[databaseName].setCollectionSchema(
+      collectionName,
+      collectionSchema
+    );
+  }
+
+  toTypescriptTypeDefinition(): string {
+    const databaseProperties = [...this.databaseNames.values()].map(
+      (databaseName) => {
+        const def = this.databaseSchemas[databaseName]
+          ? this.databaseSchemas[databaseName].toTypescriptTypeDefinition()
+          : `{}`;
+        return `      '${databaseName}': ShellAPI.Database & ${def}`;
+      }
+    );
+
+    return `{
+  ${databaseProperties.join('\n')}
+}`;
+  }
+}
+
+type AutocompleteOptions = {
+  connectionId: string;
+  databaseName: string;
+  position?: number;
+};
+
+export default class MongoDBAutocompleter {
+  private readonly context: AutocompletionContext;
+  private connectionSchemas: Record<string, ConnectionSchema>;
+  private readonly autocompleter: Autocompleter;
+
+  constructor({ context, autocompleterOptions }: MongoDBAutocompleterOptions) {
+    this.context = context;
+    this.autocompleter = new Autocompleter(autocompleterOptions);
+
+    this.connectionSchemas = Object.create(null);
+    this.connectionSchemas;
+
+    this.autocompleter.updateCode(autocompleteTypes);
+  }
+
+  addConnection(connectionId: string): ConnectionSchema {
+    if (!this.connectionSchemas[connectionId]) {
+      this.connectionSchemas[connectionId] = new ConnectionSchema();
+    }
+    return this.connectionSchemas[connectionId];
+  }
+
+  getConnectionCode(connectionKey: string): string {
+    return `
+import * as ShellAPI from '/shell-api.ts';
+import * as bson from '/bson.ts';
+
+export type ServerSchema = ${this.connectionSchemas[
+      connectionKey
+    ].toTypescriptTypeDefinition()};
+`;
+  }
+
+  getCurrentGlobalsCode(connectionId: string, databaseName: string) {
+    return `
+import { ServerSchema } from '/${connectionId}.ts';
+
+type CurrentDatabaseSchema = ServerSchema['${databaseName}'];
+
+declare global {
+  var db: CurrentDatabaseSchema;
+  var use: (collection: string) => CurrentDatabaseSchema;
```

Comment:
Hm... is this the `use db` function or something else? Why does it take a collection and return schema?